### PR TITLE
Pr sql fail fast

### PR DESCRIPTION
**Pull request by Zwelf**
Remaining changes for #4424. Improving sql write times during shutdown and slow mysql responses/timeouts. I still want to test this change, especially for the last commit. Might have time for testing this evening. Thoughts and feedback on this patchset welcome.

## Checklist

- [ ] Tested the change ingame
   - [x] in mysql mod
   - [ ] in sqlite3 mode
- [ ] Provided screenshots if it is a visual change
- [ ] Tested in combination with possibly related configuration options
- [ ] Written a unit test if it works standalone, system.c especially
- [ ] Considered possible null pointers and out of bounds array indexing
- [x] Changed no physics that affect existing maps
- [ ] Tested the change with [ASan+UBSan or valgrind's memcheck](https://github.com/ddnet/ddnet/#using-addresssanitizer--undefinedbehavioursanitizer-or-valgrinds-memcheck) (optional)

### CODE REVIEW

File: src/base/tl/threading.h [hunk 1]
```diff
@@ -2,17 +2,30 @@
...
+	// deprecated on macOS: https://stackoverflow.com/a/16655541
+	std::atomic_int m_Count{0};
 
 public:
 	CSemaphore() { sphore_init(&m_Sem); }
 	~CSemaphore() { sphore_destroy(&m_Sem); }
 	CSemaphore(const CSemaphore &) = delete;
-	void Wait() { sphore_wait(&m_Sem); }
-	void Signal() { sphore_signal(&m_Sem); }
+	int GetValue() { return m_Count.load(); }
```
Comment (heinrich5991) [inline src/base/tl/threading.h:18]: Maybe `GetApproximateValue`? It's not synchronized with anything.

**Comment by heinrich5991** [comment]
Can you explain what this does?

**Comment by Zwelf** [comment]
> Can you explain what this does?

It goes through all outstanding queries as soon as one query fails with an error (and skips read queries like `/rank`, `/mapinfo`). So that in the ddos situation only one instead of each individual query has to time out (take 60s for read and 120s for write queries). The timeout is reduced to 10s/20s in this pr as well.

When all queries are worked through, the next query will be tried at the mysql server again.

And it additionally speeds up shutdown by writing to the sqlite file only (and skipping read queries), because under normal load, there shouldn't be any sql queries left when the server shutdowns (especially with `.

I tried to additionally figure out when the queries get worked through slower than they get in, but no query fails. But I think I'll remove that commit for because I'm not so happy with my heuristic for it right now. (Current heuristic: When the expected time for the remaining queries is 60s). Another heuristic, which could be used is that one query took longer than 60s to complete (including the time in the queue).

I'll try to add some more explanatory comments in the code.

> Can we test this in an automated manner?

Can't think of a way to do this right now, but I agree that it would be good to do so :/

**Comment by heinrich5991** [comment]
bors r+

Now I feel like I understand it. :)

**Pull request merged**